libvast/vast/synopsis.hpp

```diff
@@ -0,0 +1,110 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <caf/intrusive_ptr.hpp>
+#include <caf/ref_counted.hpp>
+
+#include "vast/operator.hpp"
+#include "vast/synopsis.hpp"
+#include "vast/view.hpp"
+
+#include "vast/detail/assert.hpp"
+
+namespace vast {
+
+/// The abstract base class for synopsis data structures.
+class synopsis : public caf::ref_counted {
+public:
+  virtual ~synopsis();
+
+  /// Adds data from a table slice.
+  /// @param slice The table slice to process.
+  virtual void add(data_view x) = 0;
+
+  /// Tests whether a predicate matches. The synopsis is implicitly the LHS of
+  /// the predicate.
+  /// @param op The operator of the predicate.
+  /// @param rhs The RHS of the predicate.
+  /// @returns The evaluation result of `*this op rhs`.
+  virtual bool lookup(relational_operator op, data_view rhs) const = 0;
+};
```

Comment:
@Neverlord I'm pondering how to make this class serializable. Create a handle type derived from `ptr_handle`?

Reply:
Does anything speak against a simpler version that relies on overloads for the intrusive_ptr? For example:

```cpp
using synopsis_ptr = caf::intrusive_ptr<synopsis>;
caf::error inspect(caf::serializer& sink, synopsis_ptr& ptr);
```

Reply:
I went with the `handle` approach mostly for semantics and establishing a non-null guarantee. If you can and want deal with `nullptr` then `synopsis_ptr` is an option.